test/slam_manifold_preintegration_update_test.cc

```diff
@@ -0,0 +1,303 @@
+
+  for (int col = 0; col < 9; col++) {
+    for (int row = 0; row < 9; row++) {
+      const double covar_rc = covar(row, col);
+      const double difference = covar_rc - calculated_covar(row, col);
+      // NOTE(brad): trade-off between sample count (time) and accuracy of calculated_covar
+      CHECK(difference * difference < 9e-3 * covar_rc * covar_rc);
+    }
+  }
+}
+
+namespace example {
+static const Eigen::Vector3d accel_bias = {3.4, 1.6, -5.9};
+static const Eigen::Vector3d gyro_bias = {1.2, -2.4, 0.5};
+static const Eigen::Matrix3d accel_cov =
+    (Eigen::Matrix3d() << 7e-5, 1e-7, 1e-7, 1e-7, 7e-5, 1e-7, 1e-7, 1e-7, 7e-5).finished();
+static const Eigen::Matrix3d gyro_cov =
+    (Eigen::Matrix3d() << 1e-3, 1e-5, 1e-5, 1e-5, 1e-3, 1e-5, 1e-5, 1e-5, 1e-3).finished();
+static const Eigen::Vector3d true_gyro = Eigen::Vector3d::Constant(1.2);
+static const Eigen::Vector3d true_accel = Eigen::Vector3d::Constant(4.3);
+}  // namespace example
+
+TEST_CASE("Test ImuPreintegator.covariance", "[slam]") {
+  // In order to test that the computed covariance is correct, we sample the values many
+  // time to numerically calculate the covariance to compare.
+  using M99 = Eigen::Matrix<double, 9, 9>;
+  // Parameters
+  const double dt = 1e-4;
+  const int measurements_per_sample = 130;
+  const int sample_count = 1 << 20;
+
+  // Multithreading
+  const int thread_count = 16;
+  const int steps_per_thread = sample_count / thread_count;
+  // NOTE(brad): I assume I can equally distribute samples among the threads
+  CHECK(sample_count % thread_count == 0);
+  std::vector<std::thread> threads;
+  std::array<Eigen::Matrix<double, 9, 9>, thread_count> covariance_sums;
+  covariance_sums.fill(M99::Zero());
+
+  Eigen::MatrixXd samples(9, sample_count);
+
+  for (int i = 0; i < thread_count; i++) {
+    threads.emplace_back([&samples, &covariance_sums, dt, i]() {
+      std::mt19937 gen(1914 + i);
+      // Each thread needs its own distribution as they are not thread safe
+      MultiVarNormalDist accel_dist(example::accel_cov / dt);
+      MultiVarNormalDist gyro_dist(example::gyro_cov / dt);
+
+      Eigen::MatrixXd accel_noise;
+      Eigen::MatrixXd gyro_noise;
+
+      for (int j = steps_per_thread * i; j < steps_per_thread * (i + 1); j++) {
+        sym::ImuPreintegrator<double> integrator(example::accel_bias, example::gyro_bias);
+
+        gyro_noise = gyro_dist.Sample(gen, measurements_per_sample);
+        accel_noise = accel_dist.Sample(gen, measurements_per_sample);
+
+        for (int k = 0; k < measurements_per_sample; k++) {
+          integrator.IntegrateMeasurement(example::true_accel + accel_noise.col(k),
+                                          example::true_gyro + gyro_noise.col(k),
+                                          example::accel_cov, example::gyro_cov, dt, 0.0);
+        }
+
+        samples.col(j).segment(0, 3) =
+            integrator.PreintegratedMeasurements().DR.ToTangent(sym::kDefaultEpsilond);
```

Comment:
Oh. The explains a lot. It should be the `LocalCoordinates` w.r.t. something. Without it, what I am sampling is the correct value conjugated by:
![image](https://user-images.githubusercontent.com/84410250/214721678-30a48206-9bed-4bac-89ff-b25f7db87b82.png)
where `phi` is the noiseless value of `DR`.

Sure enough, when I cranked up the magnitude of the gyro measurements, so that this perturbation would be larger, the test started failing, and was fixed when I took the local coordinates w.r.t. the noiseless `DR`. 

I think what it should be is the covariance w.r.t. is the noiseless `DR` (in case that's different from the mean) (though I'd have to spend more time thinking about that in order to justify that).

Anyway, this demonstrates that I should use large gyroscope measurements to expose errors that might scale with `DR`.

Reply:
> I think what it should be is the covariance w.r.t. is the noiseless DR (in case that's different from the mean) (though I'd have to spend more time thinking about that in order to justify that).

Yep this makes sense to me, especially under the general assumption here that the propagating through the linearization of the update function is a good approximation